File: 207-course-schedule/207-course-schedule.py
class Solution:
    def canFinish(self, n: int, prerequisites: List[List[int]]) -> bool:
        #build graph
        graph = defaultdict(list)
        indegree = [0]* n
        for u, v in prerequisites:
            graph[v].append(u)
            indegree[u] += 1
        q = deque()
        for i in range(n):
            if indegree[i] == 0:
                q.append(i)
        cnt = 0
        while q:
            curr = q.popleft()
            cnt += 1
            for nei in graph[curr]:
                indegree[nei] -= 1
                if indegree[nei] == 0:
                    q.append(nei)
        return cnt == n

File: 207-course-schedule/207-course-schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        //bfs topological sort
        //建图
        Map<Integer, List<Integer>> graph = new HashMap<>();
        int[] inDegree = new int[numCourses];
        //建图+创建入度
        for(int[] course : prerequisites) {
            int end = course[0], start = course[1];
            graph.computeIfAbsent(start, x -> new ArrayList<>()).add(end);
            inDegree[end]++;
        }
        //将入度为零的source放入q, 进行bfs
        Queue<Integer> q = new LinkedList<>();
        
        for(int i = 0; i < numCourses; i++) { //index是node,indegree里面存的node的入度，不要搞混
            if(inDegree[i] == 0)
                q.offer(i);
        }
        int cnt = 0; //cnt 来记录入度为零的点
        while(!q.isEmpty()) {
            int curr = q.poll();
            cnt++;
            if(graph.containsKey(curr)) {
                for(int nei : graph.get(curr)) {
                    inDegree[nei]--;
                    if(inDegree[nei] == 0)
                        q.offer(nei);
                }
            }
        }
        return cnt == numCourses;
    }
}